File: w2/pset2/readability/readability.c
// readability.c
// A program which gives the reading level of text in US Grade level
// Somaditya Sinha
// 2nd January 2020

#include <stdio.h>
#include <stdlib.h>
#include <cs50.h>
#include <math.h>
#include <string.h>

int letter_count(string text);
int word_count(string text);
int sent_count(string text);

// MAIN
int main(void)
{
    string text = get_string("Text: ");
    int w = word_count(text);
    int s = sent_count(text);
    int le = letter_count(text);

    float L = (le * 100) / (float)w; // Letters per 100 words
    float S = (s * 100) / (float)w; // sentences per 100 words
    float raw_index = (0.0588 * L) - (0.296 * S) - 15.8;

    int index = round(raw_index);

    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }

    return 0;
}

// Functions
// LETTER COUNT
int letter_count(string text)
{
    int i = 0, count = 0;
    while (strcmp(&text[i], "\0"))
    {
        // A - Z
        if ((int)text[i] >= 65 && (int)text[i] <= 90)
        {
            count++;
        }
        // a - z
        if ((int)text[i] >= 97 && (int)text[i] <= 122)
        {
            count++;
        }
        i++;
    }
    return count;
}
// WORD COUNT
int word_count(string text)
{
    int i = 0, count = 0;
    while (strcmp(&text[i], "\0"))
    {
        int before = i - 1;
        if ((int)text[i] == 32)
//                         Space
        {
            if ((int)text[before] != 32)
//                                  Space
            {
                count++;
            }
        }
        if (i == strlen(text) - 1)
        {
            count++;
        }
        i++;
    }
    return count;
}

//SENTENCES COUNT
int sent_count(string text)
{
    int i = 0, count = 0;
    while (strcmp(&text[i], "\0"))
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            if (text[i - 1] != '.' || text[i - 1] != '!' || text[i - 1] != '?')
            {
                count++;
            }
        }
        i++;
    }
    return count;
}

File: w6/pset6/readability.py
from cs50 import get_string
from re import sub


def main():
    text = get_string("Text: ")
    l = letters(text)
    w = words(text)
    s = sen(text)

    print(l, w, s)

    L = (l * 100) / w
    S = (s * 100) / w

    coleLiau = round((0.0588 * L) - (0.296 * S) - 15.8)
    if coleLiau >= 16:
        print("Grade 16+")
    elif coleLiau < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {coleLiau}")


def letters(text):
    ls = 0
    for c in text:
        if (c.isalpha()):
            ls += 1
    return ls


def words(text):
    count = 1
    for c in (sub(" +", " ", text)):
        if c is " ":
            count += 1
    return count


def sen(text):
    count = 0
    for c in (sub("\.+", ".", text)):
        if c is "." or c is "?" or c is "!":
            print(c)
            count += 1
            print(count)
    return count


main()